KAFKA-8448: Too many kafka.log.Log instances (Memory Leak)

jolshan opened the pull request:
Added a method to delete the PeriodicProducerExpirationCheck task that kept adding to the log.
Includes a test to show that the task exists and then is removed when the log is closed.

Commit:
Now removes the "PeriodicProducerExpirationCheck" when a topic is deleted

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -753,6 +753,7 @@ class Log(@volatile var dir: File,
     debug("Closing log")
     lock synchronized {
       checkIfMemoryMappedBufferClosed()
+      scheduler.removeTask(task)
```
Did we consider `task.cancel` instead of this approach?

ijuma replied:
We would also have to call `setRemoveOnCancelPolicy(true)` on the executor, but that seems like a good idea anyway.

jolshan replied:
I didn't see this on the first time around. I can use this method instead.

ijuma reviewed (commented):
Thanks for the PR, one question below.

jolshan commented:
retest this please

jolshan committed:
Changed so that the task is canceled through its own method.
Needed to change the trait to return Future instead of ScheduledFuture from schedule to deal with a quirk in MockTask.

jolshan committed:
Modified so schedule returns ScheduledFuture

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -753,6 +753,7 @@ class Log(@volatile var dir: File,
     debug("Closing log")
     lock synchronized {
       checkIfMemoryMappedBufferClosed()
+        producerExpireCheck.cancel(true)
```
Nit: the indenting seems a bit off here.

ijuma commented on core/src/main/scala/kafka/utils/KafkaScheduler.scala:
```diff
@@ -50,9 +50,10 @@ trait Scheduler {
+   * @return the task scheduled
    */
-  def schedule(name: String, fun: ()=>Unit, delay: Long = 0, period: Long = -1, unit: TimeUnit = TimeUnit.MILLISECONDS)
+  def schedule(name: String, fun: ()=>Unit, delay: Long = 0, period: Long = -1, unit: TimeUnit = TimeUnit.MILLISECONDS) : ScheduledFuture[_]
```
Nit: there should be no space before `:`. Same a few lines below.

ijuma commented on core/src/main/scala/kafka/utils/KafkaScheduler.scala:
```diff
@@ -125,6 +127,12 @@ class KafkaScheduler(val threads: Int,
     }
   }
 
+  def taskRunning(task: ScheduledFuture[_]) : Boolean = {
```
Maybe we should make this `private[utils]` and mention in the comment that it's package private for testing. Also, do we actually need the `synchronnized`?

jolshan replied:
I had it there because isStarted was also synchronized, but I removed the synchronized

jolshan committed:
Modified to work with earlier versions of scala

jolshan committed:
Minor fixes

bob-barrett commented on core/src/test/scala/unit/kafka/utils/MockScheduler.scala:
```diff
@@ -96,4 +98,42 @@ case class MockTask(name: String, fun: () => Unit, var nextExecution: Long, peri
+}
+object MockTask {
+  implicit def MockTaskOrdering : Ordering[MockTask] = new Ordering[MockTask] {
+    def compare(x: MockTask, y: MockTask): Int ={
```
nit: add a space after the `=` here

bob-barrett reviewed (approved):
One minor comment, otherwise LGTM

jolshan committed:
Nit fix

jolshan commented:
retest this please

cmccabe commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -753,6 +753,7 @@ class Log(@volatile var dir: File,
     debug("Closing log")
     lock synchronized {
       checkIfMemoryMappedBufferClosed()
+      producerExpireCheck.cancel(true)
```
you should also wait until the executor has finished running, with something like `ExecutorService#awaitTermination`

ijuma replied:
I don't think you can do that. The executor is shared with other components.

jolshan replied:
So should I keep as is or change it?

cmccabe replied:
@ijuma: Good point.  We don't want to wait for the executor to terminate here.

cmccabe commented on core/src/main/scala/kafka/utils/KafkaScheduler.scala:
```diff
@@ -50,9 +50,10 @@ trait Scheduler {
    * @param period The period with which to execute the task. If < 0 the task will execute only once.
-   * @param unit The unit for the preceding times.
+   * @param unit The unit for the preceding times
+   * @return the task scheduled
```
nit: it's not really the task that was scheduled, but just a future that can be used to manage the task scheduled.

cmccabe commented on core/src/test/scala/unit/kafka/utils/MockScheduler.scala:
```diff
@@ -35,21 +35,21 @@ import org.apache.kafka.common.utils.Time
  */
 class MockScheduler(val time: Time) extends Scheduler {
-  
+
```
Let's not make so many whitespace changes to this file... it is distracting when trying to read the patch.

cmccabe commented on core/src/test/scala/unit/kafka/utils/MockScheduler.scala:
```diff
@@ -96,4 +98,42 @@ case class MockTask(name: String, fun: () => Unit, var nextExecution: Long, peri
     else
       1
   }
+
```
If we aren't going to implement the Future stuff, we should just return null for the future, so that tests relying on this behavior break quickly rather than getting weird behavior.  Alternately, it might not be that hard to properly implement it.

jolshan replied:
I tried to return null, but unfortunately it broke tests which is why I half-implemented it here.

cmccabe replied:
How did the tests break?

jolshan replied:
Any test that uses a MockTask that closes will try to call cancel on the task. This results in trying to call cancel on a null object.

cmccabe replied:
After looking at it more, I agree that returning null wouldn't work here.  However, I think returning a useless future that always appears to be successful isn't good.  Why not change the return type to Future rather than ScheduledFuture?  Then the mock classes can return a CompletableFuture which we can complete whenever we need to.

jolshan replied:
Done.

cmccabe replied:
To be clear, I don't think `MockTask` should inherit from `CompletableFuture`.  I was just suggesting that we could return a separate future object which could then do what we expected.  I think the minimum here would be cancelling the task when `future.cancel()` was called, blocking util the task was done when `future.get()` was called, etc.

If that seems too hard, we can do it in a follow on change.  If you want to get this in now, you can just go back to returning a fake future that's always completed, and we can tackle the rest in a follow-on change.

jolshan replied:
I'm a bit confused by what you mean by a separate future object that can cancel the task when future.cancel() is called.  Currently, MockScheduler creates MockTasks. Are you saying that in the schedule method, a MockTask should be created and added to the queue and then a separate future should be created that is linked to the MockTask in some way that allows it to remove the task from the queue when future.cancel() is called, etc?

 I didn't implement this functionality because MockTasks are used mostly for testing purposes, and currently there is no test to see if a MockTask can be cancelled. Perhaps we want to do that in the future, but that seems like a lot of extra work for something that is not used currently.

jolshan committed:
Fixed comment to be more accurate

jolshan commented:
retest this please

jolshan committed:
Changed to return a CompletableFuture from MockScheduler rather than ScheduledFuture

jolshan committed:
Style fixes

jolshan committed:
Modified schedule to return ScheduledFuture

cmccabe commented:
test failure is not related.  LGTM
